### ISSUE TITLE
[REVIEW]: DynamicalBilliards.jl : An easy-to-use, modular and extendable Julia package for Dynamical Billiard systems in two dimensions.

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#458]
**Submitting author:** @Datseris (<a href="http://orcid.org/0000-0002-6427-2385">George Datserus</a>)
**Repository:** <a href="https://github.com/JuliaDynamics/DynamicalBilliards.jl" target ="_blank">https://github.com/JuliaDynamics/DynamicalBilliards.jl</a>
**Version:** v1.6.1
**Editor:** @kyleniemeyer
**Reviewer:** @ahwillia
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.1059092" target="_blank">10.5281/zenodo.1059092</a>

## Status

[![status](http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd/status.svg)](http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd"><img src="http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd/status.svg)](http://joss.theoj.org/papers/753469f6b18c9c38127a7727d13c87cd)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/JuliaDynamics/DynamicalBilliards.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer questions

@ahwillia, please carry out your review in this issue by updating the checklist below (please make sure you're logged in to GitHub). The reviewer guidelines are available here: http://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @kyleniemeyer know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](http://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JuliaDynamics/DynamicalBilliards.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.6.1)?
- [x] **Authorship:** Has the submitting author (@Datseris) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @ahwillia it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As as reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by kyleniemeyer** [comment]
👋 @Datseris @ahwillia @mlxd

**Comment by Datseris** [comment]
Thank you very much for accepting to review my submission.

**Comment by mlxd** [comment]
Happy to participate. It will likely be the coming weekend before I can spend time going through everything in detail. As suggested by @kyleniemeyer I will leave comments to the above check-box items.

**Comment by ahwillia** [comment]
Overall this looks excellent - really appreciate your attention to detail while putting this together!

@Datseris - what do you think about having an `examples/` folder containing jupyter notebooks? These could just contain the code that you have under `Tutorials` in the docs (but without the long text). That way people can flip back and forth between reading your docs and trying it. Right now there is an awkward copy-paste step.

Other than that I'm inclined to approve this. But we can wait for @mlxd to weigh in.

**Comment by ahwillia** [comment]
Oh two other short comments

>> **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

Is this listed anywhere explicitly?

>> **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

Do you mind adding DOI links to the references in your `paper.md` document? Thanks!

**Comment by Datseris** [comment]
Thanks for the kind words @ahwillia . I've just added guidelines for issue reports. 

I will now do your other requests:
* Add DOIS to paper.md
* Add a dedicated paragraph on the homepage of the documentation about support (we have our own chatroom on gitter)
* Add a dedicated contributors_guide file either in the repo first page or in the documentation page.

About the Jupyter notebooks... Is it possible to avoid this step? My only issue is that I have never used Jupyter and it make take a while/be tricky to set it up etc.

I know it is kind of annoying to copy/paste that much but fortunately the documentation allows you to copy an entire code-block by clicking the top - right corner of each code block.

**Comment by ahwillia** [comment]
Yep - I'm fine with no jupyter notebooks, just a suggestion.

**Comment by Datseris** [comment]
@ahwillia I have done the requested changes. Now all citations of the `paper.md` have DOIs, and the homepage has 2 sections: One for support and one for contributing.

See this link: https://juliadynamics.github.io/DynamicalBilliards.jl/latest/#support

Of course, the changes apply only on the latest version, and not on the release tag 1.6.1. Should a new tag be required let me know and I will tag it.

**Comment by ahwillia** [comment]
I'm happy to accept this - I was able to install, run tests, and reproduce some simple plots. The package supports a well-defined need and is reasonably mature. Looking forward to seeing future developments!

Now we'll just wait on @mlxd to check it over.

**Comment by mlxd** [comment]
Everything is clearly documented, installation and use is as described. This is a very useful package, and I'd be happy to accept this too. I agree with all of @ahwillia's comments and checkboxes.

I do have one suggestion on the documentation for the examples:
For the Julia-logo Billiard example (and potentially elsewhere) [https://juliadynamics.github.io/DynamicalBilliards.jl/stable/tutorials/examples/], the file save path seems Windows specific (sname = "C:\\***\\anim"). 

While it works for me on Mac, it does create a log of "C:\***\anim_99.png" files in the current directory, which may go elsewhere on a Windows machine due to the path (I have not tested). I would suggest a more platform-independent save string with the correct format for Win/Mac/LNX. Maybe explicitly setting the savedir as the current working directory (pwd()) would suffice for all platforms, and users can run the example code without changing. Though, feel free to ignore this, as it is merely a suggestion.

**Comment by Datseris** [comment]
This was simply a "dummy" name (indicated by the ***)... I expected users to put as a string wherever they want to save their files. 

I will just replace `pwd()` since you are right, it is better.

**Comment by mlxd** [comment]
Great. Excellent work on this. It was very easy to review and recommend for acceptance. Unless @ahwillia has anything else, I think we are done.

**Comment by Datseris** [comment]
Awesome. Thank you for the swift and concise review!

**Comment by kyleniemeyer** [comment]
@Datseris sounds like we are nearly good to go—let me know when you make the change mentioned by @mlxd above.

**Comment by Datseris** [comment]
@kyleniemeyer the change is already done: https://juliadynamics.github.io/DynamicalBilliards.jl/latest/#julia-billiard-animation

**Comment by kyleniemeyer** [comment]
@Datseris ok, there are a few changes required for the paper.

1. The references need to be extracted from the Markdown paper file, put in a BibTeX .bib file, and then cited in the paper. You can see an example in the JOSS [author guidelines](http://joss.theoj.org/about#author_guidelines). This is necessary for our LaTeX-based compiling of the paper PDF, and also depositing the citation info in CrossRef.

2. The animated GIF you have is nice for the README, but it won't work for the paper, which is generated as a static PDF. Instead of the logo, could you perhaps include an example figure using results produced by the package (perhaps from one of the examples/tutorials), along with a brief description? That would be more useful and informative.

3. As for the actual content of the paper, the layout you have is not quite appropriate—again, it works for a README-type document, but not the paper. Links to your software repository will already be included in the paper (see [an example](https://github.com/openjournals/joss-papers/blob/joss.00460/10.21105.joss.00460.pdf) compiled PDF for a different paper under review), so please remove those from the top.

Also, to the summary or introductory paragraph, could you describe the research application of the software for a diverse, non-specialist audience? Many people may not be familiar with billiard systems in the context of research.

4. A few minor typos/suggested edits:
- "In [2] this is described perfectly, where Bunimovich and Vela-Arevalo (two pioneers in the field) summarize new insights in the field of dynamical billiards" -> "Bunimovich and Vela-Arevalo (two pioneers in the field) summarized new insights in the field of dynamical billiards [2]."
- "Our package has plenty of applications" is quite informal and not descriptive
- "which results to accuracies" -> "which results in accuracies"
- I would eliminate the use of italics to emphasize words throughout; it makes the paper read as more persuasive than is intended/appropriate, I think.

**Comment by Datseris** [comment]
@kyleniemeyer I have done all the changes you requested. Please see the latest files here: https://github.com/JuliaDynamics/DynamicalBilliards.jl/tree/master/paper

I have a problem though. You said that "links to the repository will be included in the paper". I saw the example you cited but nowhere in that example there was a link to a *documentation page*. There was a link to a GitHub repository, but for me this is not enough.

**I would like to have a direct link to the documentation of the package somewhere in the paper**. This could be either at the side, as part of this typical information or in-text. I do not mind.

The users of a package are users, not developers. They should never have to know what git or GitHub is. To use the package you should just have a link to the documentation without having to go through links to other links to search for a documentation page.

**Comment by kyleniemeyer** [comment]
@Datseris sure, that's fine—perhaps you could link on "detailed documentation archive" in the summary at the top of the paper?

**Comment by Datseris** [comment]
@kyleniemeyer I am glad we could come to an agreement. I have added an in-text hyperlink.

Is that all from my side?

**Comment by kyleniemeyer** [comment]
@Datseris thanks! I'm going to see if the paper compiles. The last step from you is archiving the final version of the software etc. (like using Zenodo) and then telling me the DOI here; but, wait until we're sure the paper doesn't need any more changes.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by kyleniemeyer** [comment]
@arfon ruh roh. any way to see what went wrong?

**Comment by arfon** [comment]
Yeah, I'm debugging now. Getting errors reported back in here is somewhat tricky.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
PDF failed to compile for issue #458 with the following error: 

 pandoc-citeproc: Could not find none
CallStack (from HasCallStack):
  error, called at src/Text/CSL/Pandoc.hs:224:39 in pandoc-citeproc-0.10.4-BdOyQb33rzG2TMOLj4Fbp9:Text.CSL.Pandoc
pandoc: Error running filter pandoc-citeproc
Filter returned error status 1
Looks like we failed to compile the PDF

```

**Comment by arfon** [comment]
@Datseris - looks like you need to link to the `paper.bib` file. You currently have this set as `bibliography: none`

**Comment by whedon** [comment]
```
PDF failed to compile for issue #458 with the following error: 

 pandoc-citeproc: Could not find none
CallStack (from HasCallStack):
  error, called at src/Text/CSL/Pandoc.hs:224:39 in pandoc-citeproc-0.10.4-BdOyQb33rzG2TMOLj4Fbp9:Text.CSL.Pandoc
pandoc: Error running filter pandoc-citeproc
Filter returned error status 1
Looks like we failed to compile the PDF

```

**Comment by Datseris** [comment]
@arfon I have changed it to `bibliography: paper.bib`. I hope this is enough.

**Comment by Datseris** [comment]
@kyleniemeyer I actually do not know what you mean with:
>The last step from you is archiving the final version of the software etc. (like using Zenodo) and then telling me the DOI here; 

Isn't the DOI the duty of the Journal and not the author? I also do not know what "Zenodo" is.

**Comment by kyleniemeyer** [comment]
@Datseris we provide a DOI associated with the paper, but the author is responsible for archiving the accepted version of the software (i.e., your repository on GitHub). We do not archive the software and associated material, just the paper. You can see an example in a [recently published paper](http://www.theoj.org/joss-papers/joss.00451/10.21105.joss.00451.pdf) with the "Archive" link to the left.

This is from the JOSS About page:
> Upon successful completion of the review, deposit a copy of your (updated) software repository with a data-archiving service such as Zenodo or figshare, issue a DOI for the software archive, and update the review issue thread with your DOI.

I recommend using [Zenodo](https://zenodo.org), but [Figshare](https://figshare.com) is another option. Both are free, and used commonly for this purpose.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
PDF failed to compile for issue #458 with the following error: 

 pandoc-citeproc: Could not find none
CallStack (from HasCallStack):
  error, called at src/Text/CSL/Pandoc.hs:224:39 in pandoc-citeproc-0.10.4-BdOyQb33rzG2TMOLj4Fbp9:Text.CSL.Pandoc
pandoc: Error running filter pandoc-citeproc
Filter returned error status 1
Looks like we failed to compile the PDF

```

**Comment by Datseris** [comment]
@kyleniemeyer okay I see. Seems simple enough. Let me know when `whedon` can compile the pdf and then I will do it. Or should I do it now?

**Comment by kyleniemeyer** [comment]
Let's wait to see if any other changes are needed for the paper first, since it looks there this is still some problem.

@arfon it looks like whedon is raising the same error, but I can see in the `paper.md` that the bibliography seems set correctly now.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
https://github.com/openjournals/joss-papers/blob/joss.00458/10.21105.joss.00458.pdf
```

**Comment by Datseris** [comment]
@arfon The compiled pdf does not have a a hyperlink to the documentation page.

**Comment by kyleniemeyer** [comment]
@Datseris the link is actually present if you download the PDF, but the text is just not colored blue.

@arfon is there perhaps a `hyperref` setting for `href` links that is needed in the LaTeX template?

**Comment by kyleniemeyer** [comment]
(Sorry for the paper gen hiccups @Datseris, we are still testing out our new automated paper compilation)

**Comment by Datseris** [comment]
No problem! I also apologize because I did not realise that there was a link there. If you can add the standard hyperref color that would be much appreciated. If in the end of the day it appears not to be possible, then I can live with it being a "hidden link".

As a side note: how soon after the pdf generation works can I get a working DOI/reference to your journal? I am going to submit a publication (to another journal) next week and I am thinking if I can wait so I can include an official JOSS reference instead of a reference to the documentation page in my paper.

**Comment by kyleniemeyer** [comment]
@Datseris you will have the JOSS DOI basically as soon as we sort out this paper issue, so you should be able to include the full published JOSS reference.

**Comment by Datseris** [comment]
That is excellent. Since the paper seems to be in a final form besides this small hyperref detail, should I go ahead and perform this "Zenodo" process you described?

**Comment by kyleniemeyer** [comment]
Actually, I caught one final typo fix to make first:
change 
> The package is mainly used to simulate any kind of 2 dimensional system where particle motion is interrupted by collisions, like e.g. a gas infused with big molecules. 

to 

>The package is mainly used to simulate any kind of two-dimensional system where particle motion is interrupted by collisions, e.g., a gas infused with big molecules.

Once you make that fix, please go ahead and archive the repository, and then report the DOI here. On Zenodo, you can do this manually by uploading a .zip (e.g.) of the repository; alternatively, Zenodo has a [nice GitHub integration](https://guides.github.com/activities/citable-code/) where it will automatically grab everything if you make a release on GitHub. (I do this for most of my own software packages.)

**Comment by Datseris** [comment]
Wow, the process was extremely smooth and straightforward!
I am pasting the badge here:



[![DOI](https://zenodo.org/badge/77843944.svg)](https://zenodo.org/badge/latestdoi/77843944)

**Comment by kyleniemeyer** [comment]
@whedon set 10.5281/zenodo.1059092 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.1059092" target="_blank">10.5281/zenodo.1059092</a> is the archive.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
https://github.com/openjournals/joss-papers/blob/joss.00458/10.21105.joss.00458.pdf
```

**Comment by arfon** [comment]
@kyleniemeyer @Datseris - the links in the paper are looking a bit better now.

**Comment by kyleniemeyer** [comment]
Yes, looks great! Ready to accept/publish.

**Comment by arfon** [comment]
@ahwillia - many thanks for your review here and to @kyleniemeyer for editing this submission ✨ 

@Datseris - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00458 ⚡️ 🚀 💥

**Comment by Datseris** [comment]
Awesome! Thanks everybody for their effort!

**Comment by Datseris** [comment]
@ahwillia @kyleniemeyer I am sorry about this but is it possible to recompile the paper? Or "update" it?

I realized that even though I am acknowledging people properly in the repository page, this is not very transparent in the paper. Therefore I copied the Acknowledgements page from the repository and added it to the paper.md as well, so that they are more transparent.

I am sorry for forgetting this.

**Comment by arfon** [comment]
@Datseris - yes, that's OK. The `@username` syntax in the paper is confusing Pandoc though:

<img width="701" alt="screen shot 2017-11-20 at 6 48 23 am" src="https://user-images.githubusercontent.com/4483/33017054-eb25a93e-cdbe-11e7-9b48-265d8f2ddc31.png">

**Comment by arfon** [comment]
I think you should be able to quote their usernames in back-ticks e.g. ````@arfon````

**Comment by Datseris** [comment]
Thanks a lot for the flexibility you provide. I changed the nametags tp have back-ticks

**Comment by Datseris** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
https://github.com/openjournals/joss-papers/blob/joss.00458/joss.00458/10.21105.joss.00458.pdf
```

**Comment by Datseris** [comment]
@arfon everything is fine with the pdf! The only issue is that when I go to the DOI page,
the pdf linked there is not updated (it still has the old version): http://joss.theoj.org/papers/10.21105/joss.00458

**Comment by arfon** [comment]
> @arfon everything is fine with the pdf! The only issue is that when I go to the DOI page,
the pdf linked there is not updated (it still has the old version): http://joss.theoj.org/papers/10.21105/joss.00458

Yep, I've not updated them yet. Will get to this either later today or tomorrow.

**Comment by arfon** [comment]
> @arfon everything is fine with the pdf! The only issue is that when I go to the DOI page,
the pdf linked there is not updated (it still has the old version): http://joss.theoj.org/papers/10.21105/joss.00458

@Datseris - ok, these should be updated now.

**Comment by Datseris** [comment]
@arfon thank you very much, everything looks very nice. Quick question: what is the "formal"/official form for the bibtex entry for your journal? From my DOI link I can see:

>Citation:
Datseris, (2017). DynamicalBilliards.jl: An easy-to-use, modular and extendable Julia package for Dynamical Billiard systems in two dimensions.. Journal of Open Source Software, 2(19), 458, doi:10.21105/joss.00458

Is there any way to export the citation as a bibtex entry?

**Comment by kyleniemeyer** [comment]
@Datseris you can try using https://www.doi2bib.org/; we used to rely on it, but it isn’t always working.

**Comment by danielskatz** [comment]
also see https://github.com/openjournals/joss/issues/341
